src/NVMLIOGroup.cpp

```diff
@@ -337,60 +338,67 @@ namespace geopm
         return result;
     }
 
+    // The active process list NVML call can be costly, 0.5-2ms per call was seen in early testing on average,
+    // with a worst case of 8ms per call.  Because of this we cache the processes in a PID <-> Accelerator map
+    // before using them elsewhere
+    std::map<pid_t, int> NVMLIOGroup::accelerator_process_map(void) const
+    {
+        std::map<pid_t,int> accelerator_pid_map;
+
+        for (int accel_idx = 0; accel_idx < m_platform_topo.num_domain(GEOPM_DOMAIN_BOARD_ACCELERATOR); ++accel_idx) {
+            std::vector<int> active_process_list = m_nvml_device_pool.active_process_list(accel_idx);
+            for (auto proc_itr : active_process_list) {
+                // If a process is associated with multiple accelerators we have no good means of
+                // signaling the user beyond providing an error value (-1).
```

Comment:
Comment doesn't seem to match.  Aren't there three cases here (0, 1, >1) ?  Is the first case intended to be count == 1?

Reply:
The expected return from this function is a map of PIDs to accelerator id (the NVML Device index), not the count of accelerators.  The expected values for the map are -1 to N, where N is the number of accelerators. 

As we iterate over the active process list for each accelerator we add a key-value pair into the map, but if at any point the .count function returns that a matching key already exists in the map for this PID we set the value for that PID to -1 to indicate there isn't a single accelerator mapping that we can report (read_signal & sample require a single value to return)

The intent is that for the following PID to accelerator index values:
123, 0
124, 1
124, 2
125, 3

We will return
123: 0
124: -1
125: 3

Reply:
got it